File: 0907-sum-of-subarray-minimums/0907-sum-of-subarray-minimums.cpp
class Solution {
public:
    int sumSubarrayMins(vector<int>& arr) {
        // sort(arr.begin(),arr.end());
        // int sum=0;
        // for(int i=0;i<arr.size();++i)
        // {
        //     for(int j=i;j<arr.size();++j)
        //     {
        //         sum+=arr[i];
        //     }
        // }
        // return sum;
        
        arr.insert(begin(arr), 0);
        
        int n = size(arr), mod = 1e9 + 7, sum = 0;
        vector<int> res(n);
        stack<int> s;
        s.emplace(0);
        
        for(int i = 0; i < n; ++i) {
            while(!s.empty() and arr[s.top()] > arr[i])
                s.pop();
            int j = s.top();
            res[i] = res[j] + (i - j) * arr[i];
            s.emplace(i);
        }
        
        for(int i = 0; i < n; ++i)
            sum = (sum + res[i]) % mod;
        
        return sum;
    }
};

File: 0907-sum-of-subarray-minimums/0907-sum-of-subarray-minimums.java
class Solution {
    public int sumSubarrayMins(int[] A) {
      // array tracking previous less element
        
        Stack<int[]> previousLess = new Stack<>();
        
        int[] leftDistance = new int[A.length];
        int[] sums = new int[A.length];
        // Arrays.fill(sums, 1);
        for(int i=0; i<A.length; i++)
        {
            // use ">=" to deal with duplicate elements
            while(!previousLess.isEmpty() && previousLess.peek()[0] >= A[i])
            {
                previousLess.pop();
            }
            
            int j = previousLess.isEmpty() ? -1 : previousLess.peek()[1];
            sums[i] = previousLess.isEmpty() ? A[i] * (i+1) : sums[j] + A[i] * (i-j);
            previousLess.push(new int[]{A[i], i});
            
        }
              
        long MOD = (long) 1000000007;
        long ans = 0;

        for(int i=0; i<A.length; i++)
        {
            ans = (ans + (long)sums[i]) % MOD;
        }
        return (int)ans;
    }
}